core/uploadDiveLogsDE.cpp

```diff
@@ -40,15 +41,13 @@ void uploadDiveLogsDE::doUpload(bool selected, const QString &userid, const QStr
 
 
 	/* generate a temporary filename and create/open that file with zip_open */
-	QString filename(QDir::tempPath() + "/divelogsde-upload.dld");
-
-	// delete file if it exist
-
-	QFile f(filename);
-	if (f.open(QIODevice::ReadOnly)) {
-		f.close();
-		f.remove();
-	}
+	auto *tmpfile = new QTemporaryFile();
+	tmpfile->setFileTemplate(QDir::tempPath() + "/divelogsde-upload.XXXXXXXX.dld");
+	tmpfile->open();
+	QString filename = tmpfile->fileName();
+	tmpfile->close();
```

Comment:
This looks like almost the same race condition you are trying to resolve: What happens if the other process creates the file between you closing it here and it being reopened for writing? This only produces a filename that at some point in the past did not exist. Once you opened it above you should write the proper contents to it.

Reply:
That could only happen if both users generate the same random file name at the same time, which is highly unlikely.

From what I see elsewhere in the code, the only way to directly use the temporary file name for zip_open is using an OS- and version-specific #ifdef case distinction and dup()ing the underlying fd, which I thought might be worth avoiding.

Reply:
Ah, now that you mention it, I remember there was a problem with libzip IIRC that it cannot be used with a file handle only with a filename. You could use the PID in the template which would at least make sure you do not step on the foot of another incarnation of Subsurface on the same machine. 

I have to admit I don't know how Qt fills the pattern, if it tries a random string or some deterministic way to find the 'first' non-existent file name.

Reply:
On my system this comes out as /tmp/divelogsde-upload.mBToUtKs.dld so I assume it's random. I can add the PID if that helps, even though I'm not sure what to use because I assume getpid() is not portable (QCoreApplication::applicationPid() ?  - I'm not a Qt guy really). Another option would be to use a UUID, or even make sure each instance has its own private tmp dir (although I would consider that a more invasive change). Or perhaps leave the file name static and instead go for a dedicated temp dir via QTemporaryDir.

(Also in case this wasn't clear from my description: What this tries to solve is not a race condition. The invocations of the upload by two different users could be days apart. The second one will fail because the file (left over by the first one) would already exist but is owned by a different user and so is not deletable.)

Reply:
That function should give us a random file name based on the given template, and IMO that should be enough for us. Anything more complex sounds overkill :D